pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java

```diff
@@ -212,39 +218,132 @@ private IdealState getNextState(IdealState currentState, IdealState targetState,
+        .getInt(RebalanceUserConfigConstants.MIN_REPLICAS_TO_KEEPUP_FOR_NODOWNTIME,
+            RebalanceUserConfigConstants.DEFAULT_MIN_REPLICAS_TO_KEEPUP_FOR_NODOWNTIME);
+    LOGGER.debug("No downtime mode is enabled. Need to keep {} serving replicas up while rebalancing",
+        minReplicasToKeepUp);
+
+    if (minReplicasToKeepUp >= currentIdealStateSegmentHosts.size()) {
+      // if the minimum number of serving replicas in rebalance config is
+      // greater than or equal to number of replicas of a segment, then it
+      // is impossible to honor the request. so we use the default number
+      // of minimum serving replicas we will keep for no downtime mode
+      // (currently 1)
+      minReplicasToKeepUp = RebalanceUserConfigConstants.DEFAULT_MIN_REPLICAS_TO_KEEPUP_FOR_NODOWNTIME;
+    }
+    MapDifference difference = Maps.difference(targetIdealStateSegmentHosts, currentIdealStateSegmentHosts);
+    if (difference.entriesInCommon().size() >= minReplicasToKeepUp) {
+      // if there are enough hosts in common between current and target ideal states
+      // to satisfy the min replicas condition, then there won't be any downtime
+      // and we can directly set the current ideal state to target ideal state
+      LOGGER.debug("Current and target ideal states have common hosts. Will set to target state at one go");
+      setTargetState(idealStateToUpdate, segmentId, targetIdealStateSegmentHosts);
+      ++_rebalancerStats.directTransitions;
     } else {
       // remove one entry
-      idealState.getInstanceStateMap(segmentId).remove(srcMap.keySet().stream().findFirst().get());
+      String hostToRemove = "";
+      for (String host : currentIdealStateSegmentHosts.keySet()) {
+        // the common host between current and target ideal
+        // states should be ignored
+        if (!targetIdealStateSegmentHosts.containsKey(host)) {
+          hostToRemove = host;
+          break;
+        }
+      }
+
+      if (!hostToRemove.equals("")) {
+        idealStateToUpdate.getInstanceStateMap(segmentId).remove(hostToRemove);
+        LOGGER.info("Removing host: {} for segment: {}", hostToRemove, segmentId);
+      }
+
       // add an entry from the target state to ensure there is no downtime
-      String instanceId = targetMap.keySet().stream().findFirst().get();
-      idealState.setPartitionState(segmentId, instanceId, targetMap.get(instanceId));
-      LOGGER.debug("Adding " + instanceId + " to serve segment " + segmentId);
+      final Map<String, String> updatedSegmentInstancesMap = idealStateToUpdate.getInstanceStateMap(segmentId);
+      String hostToAdd = "";
+      for (String host : targetIdealStateSegmentHosts.keySet()) {
+        // the hosts host that has already been added from target
+        // ideal state to new state should be ignored
+        if (!updatedSegmentInstancesMap.containsKey(host)) {
+          hostToAdd = host;
+          break;
+        }
+      }
+
+      if (!hostToAdd.equals("")) {
```

Comment:
assert that it does not equal "", or, better, add a preconditions check

Reply:
done